MediaRecorder WebM Videos Play Rotated 90° Counterclockwise

jyavenard opened the pull request:
#### 83051fcd6af8b80ba5b1c2ee71e8d59828820a61
<pre>
MediaRecorder WebM Videos Play Rotated 90° Counterclockwise
<a href="https://bugs.webkit.org/show_bug.cgi?id=290223">https://bugs.webkit.org/show_bug.cgi?id=290223</a>
<a href="https://rdar.apple.com/147878206">rdar://147878206</a>

Reviewed by Youenn Fablet.

WebM, unlike mp4 doesn&apos;t support metadata indicating that a video track
should be played with given rotation.
So instead we apply the rotation on the RealtimeMediaSource before encoding it
to vp8 or vp9.

Added test. We do not test on real iOS devices, manually tested on iPhone.

* LayoutTests/fast/mediastream/mediaStream-with-rotation-expected.txt: Added.
* LayoutTests/fast/mediastream/mediaStream-with-rotation.html: Added.
* LayoutTests/platform/glib/TestExpectations: Skip test on GStreamer platform, test times out.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.cpp:
(WebCore::MediaRecorderPrivateAVFImpl::create):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::initialize):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h:
(WebCore::MediaRecorderPrivateWriter::shouldApplyVideoRotation const):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.h:

Canonical link: <a href="https://commits.webkit.org/294257@main">https://commits.webkit.org/294257@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/41f649e60299d42d7274efe5b8f98e62cdace683

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101315 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20977 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11281 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106465 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51942 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103356 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21286 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29472 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/77165 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34199 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104322 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16417 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91510 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57512 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/16235 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9522 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51290 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/86109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9596 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108819 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28443 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86140 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28805 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87717 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/85699 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21791 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30423 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8131 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22547 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28373 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/33647 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28185 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31505 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29743 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.cpp:
```diff
@@ -120,7 +120,9 @@ std::unique_ptr<MediaRecorderPrivateAVFImpl> MediaRecorderPrivateAVFImpl::create
     }
     if (selectedTracks.videoTrack) {
-        recorder->setVideoSource(&selectedTracks.videoTrack->source());
+        auto& source = selectedTracks.videoTrack->source();
```
```suggestion
        Ref source = selectedTracks.videoTrack->source();
```

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.cpp:
```diff
@@ -120,7 +120,9 @@ std::unique_ptr<MediaRecorderPrivateAVFImpl> MediaRecorderPrivateAVFImpl::create
     if (selectedTracks.videoTrack) {
-        recorder->setVideoSource(&selectedTracks.videoTrack->source());
+        auto& source = selectedTracks.videoTrack->source();
+        source.setShouldApplyRotation(recorder->shouldApplyVideoRotation());
```
I would tend to do:
```
if (recorder->shouldApplyVideoRotation())
    source.setShouldApplyRotation(true);
```
This removes a virtual function call in the mp4 case.
At some point, I'll likely remove the boolean from RealtimeMediaSource::setShouldApplyRotation.

youennf commented on LayoutTests/fast/mediastream/mediaStream-with-rotation-webm.html:
```diff
@@ -0,0 +1,46 @@
+        if (window.testRunner)
+            testRunner.setMockCameraOrientation(90);
+        const stream = await navigator.mediaDevices.getUserMedia({ video : true });
+        let result1 = await setVideo(t, stream);
```
```suggestion
        const result1 = await setVideo(t, stream);
```
Ditto for result2 and blob.

youennf commented on LayoutTests/fast/mediastream/mediaStream-with-rotation-webm.html:
```diff
@@ -0,0 +1,46 @@
+            setTimeout(reject, 5000);
+        });
+        recorder.start();
+        setTimeout(() => recorder.stop(), 1500);
```
Do we need to wait 1500 seconds?
Can we just do a start(100) and stop after the first non empty blob or maybe two non empty blobs?

youennf commented on LayoutTests/fast/mediastream/mediaStream-with-rotation-webm.html:
```diff
@@ -0,0 +1,46 @@
+        const stream = await navigator.mediaDevices.getUserMedia({ video : true });
+        let result1 = await setVideo(t, stream);
+
+        let blob = await record(stream, 'video/webm');
```
Since we are it, can we also add a `video/mp4` test if it is not already the case?

jyavenard committed:
MediaRecorder WebM Videos Play Rotated 90° Counterclockwise
https://bugs.webkit.org/show_bug.cgi?id=290223
rdar://147878206

Reviewed by Youenn Fablet.

WebM, unlike mp4 doesn't support metadata indicating that a video track
should be played with given rotation.
So instead we apply the rotation on the RealtimeMediaSource before encoding it
to vp8 or vp9.

Added test. We do not test on real iOS devices, manually tested on iPhone.

* LayoutTests/fast/mediastream/mediaStream-with-rotation-expected.txt: Added.
* LayoutTests/fast/mediastream/mediaStream-with-rotation.html: Added.
* LayoutTests/platform/glib/TestExpectations: Skip test on GStreamer platform, test times out.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.cpp:
(WebCore::MediaRecorderPrivateAVFImpl::create):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::initialize):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h:
(WebCore::MediaRecorderPrivateWriter::shouldApplyVideoRotation const):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.h:

Canonical link: https://commits.webkit.org/294257@main
